lib/addons/velero/index.ts

```diff
@@ -0,0 +1,259 @@
+     * @default 
+     *      namespace:{
+     *        name: "velero",
+     *        create: true    
+     *      }
+     */
+    namespace?: {
+        name: string,   // default value is "velero", if it is specified then it will be the namespace of where velero is deployed to.
+        create: boolean // default is true, if it is false, no new namespace will be created
+    };
+
+     /**
+     * Values to pass to the chart as per https://github.com/vmware-tanzu/helm-charts/blob/main/charts/velero/values.yaml#
+     * Required if provided.
+     */
+    values: {
+        [key: string]: any;
+    };
+
+}
+
+/**
+ * Defaults options for the add-on
+ */
+const defaultProps: VeleroAddOnProps = {
+    version: "2.23.6",
+    namespace: {
+        name: "velero",
+        create: true
+    },
+    values:{
+        initContainers:[
+            {
+                name: "velero-plugin-for-aws",
+                image: "velero/velero-plugin-for-aws:v1.2.0",
+                imagePullPolicy: "IfNotPresent",
+                volumeMounts:[
+                    {
+                        mountPath: "/target",
+                        name: "plugins"
+                    }
+                ]
+            }
+        ],
+        configuration: {
+            provider: "aws",
+            backupStorageLocation:{
+                name: "default",
+                config:{}
+            },
+            volumeSnapshotLocation:{
+                name: "default",
+                config:{}
+            },
+        },
+        serviceAccount: {
+            server:{}
+        }
+    },
+
+};
+
+export class VeleroAddOn implements ClusterAddOn {
+
+    private options: VeleroAddOnProps;
+    private chartNode: HelmChart;
+    constructor(props?: VeleroAddOnProps) {
+        if (props) {
+            // merge the nested json files
+            this.options = merge(defaultProps, props);
+        }
+        else {
+            this.options = defaultProps
+        }
+    }
+
+    /**
+     * Implementation of the add-on contract deploy method.
+    */
+    async deploy(clusterInfo: ClusterInfo): Promise<Construct> {
+        const cluster = clusterInfo.cluster;
+        const props = this.options;
+               
+        // Create S3 bucket if no existing bucket, create s3 bucket and corresponding KMS key
+        const s3Bucket = this.getOrCreateS3Bucket(clusterInfo, "backup-bucket", props.values.configuration.backupStorageLocation.bucket)
+
+        // Create Namespace if namespace is not explicied defined.
+        const veleroNamespace = this.createNamespaceIfNeeded(clusterInfo, "velero", props.namespace);
+
+        // Setup IAM Role for Service Accounts (IRSA) for the Velero Service Account
+        const veleroServiceAccount = this.createServiceAccountWithIamRoles(clusterInfo, "velero-account", veleroNamespace, s3Bucket);
+        //veleroServiceAccount.node.addDependency(veleroNamespace);
```

Comment:
@shapirov103 Line 113 doesnt work, as in line 109 `veleroNamespace` is a string, as in this scenario it takes into account an existing namespace, in that case no namespace is required to be created. That is why i commented it out.

Reply:
However, line 112 has already had the inner dependency, as the parameter `veleroNamespace` is referred to the line 109, thus, it has to rely on namespace being created first.

```typescript
const veleroServiceAccount = this.createServiceAccountWithIamRoles(clusterInfo, "velero-account", veleroNamespace, s3Bucket);
```

Reply:
The thing about CDK is that it will produce a CloudFormation Template, and for the template the order that you defined in your code does not matter. You need to use CFN dependency mechanism to handle this.